### PR TITLE
[Beam-6858] Validate that side-input parameters match the type of the PCollectionView

### PR DESCRIPTION
**Pull request by reuvenlax**
Validate that the type of a processElement side-input parameter matches that of the PCollectionView.

Note: Suppliers are used because often the TypeDescriptor is not Serializable, and so cannot be captured in the PCollectionView.

### CODE REVIEW

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluatorFactory.java [hunk 1]
```diff
@@ -107,7 +108,8 @@ public void cleanup() throws Exception {
       List<PCollectionView<?>> sideInputs,
       TupleTag<OutputT> mainOutputTag,
       List<TupleTag<?>> additionalOutputTags,
-      DoFnSchemaInformation doFnSchemaInformation)
+      DoFnSchemaInformation doFnSchemaInformation,
+      Map<String, String> sideInputMapping)
```
Comment (salmanVD) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluatorFactory.java]: @reuvenlax This should be String, String? or PCollectionView

**Comment by reuvenlax** [comment]
Yes, we should change to Map<String, PCollectionView<?>>

**Comment by reuvenlax** [comment]
R: @TheNeuralBit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java [hunk 2]
```diff
@@ -437,6 +438,29 @@ private static void validateStateApplicableForInput(DoFn<?, ?> fn, PCollection<?
...
+    for (SideInputParameter sideInput : processElementMethod.getSideInputParameters()) {
+      PCollectionView<?> view = sideInputs.get(sideInput.sideInputId());
+      checkArgument(
+          view != null,
+          "the ProcessElement method expects a side input identified with the tag %s, but no such side input was"
+              + " supplied. Use withSideInput(String, PCollectionView) to supply this side input.",
+          sideInput.sideInputId());
+      TypeDescriptor<?> viewType = view.getViewFn().getTypeDescriptor();
+
+      // Currently check that the types exactly match, even if the types are convertible.
```
Comment (TheNeuralBit) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java]: Are there plans to handle convertible types? Should there be a JIRA thats linked here?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 3]
```diff
@@ -737,6 +737,76 @@ public void testParDoWithSideInputs() {
...
+          };
+
+      thrown.expect(IllegalArgumentException.class);
+      PCollection<List<Integer>> output =
+          pipeline
+              .apply("Create main input", Create.of(2))
+              .apply(ParDo.of(fn).withSideInput(sideInputTag1, sideInput1));
+      pipeline.run();
+    }
+
```
Comment (TheNeuralBit) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Would you mind adding tests for the remaining view types - Map, Multimap, Iterable?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java (hunk 2 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java]: We have the same issue with Element parameters. I'll check to see if we have a JIRA for this (I think a user actually filed one some time back).

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java (hunk 3 above)
Comment (reuvenlax) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: tests added

**Comment by reuvenlax** [comment]
all comments addressed

**Pull request merged**